src/input_common/joycon/jc_adapter.cpp

```diff
@@ -0,0 +1,372 @@
+// Copyright 2014 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <chrono>
+#include <thread>
+#include "common/logging/log.h"
+#include "input_common/joycon/Jc_adapter.h"
+
+namespace JCAdapter {
+
+/// Used to loop through and assign button in poller
+constexpr std::array<PadButton, 12> PadButtonArray{
+    PadButton::PAD_BUTTON_LEFT, PadButton::PAD_BUTTON_RIGHT, PadButton::PAD_BUTTON_DOWN,
+    PadButton::PAD_BUTTON_UP,   PadButton::PAD_TRIGGER_Z,    PadButton::PAD_TRIGGER_R,
+    PadButton::PAD_TRIGGER_L,   PadButton::PAD_BUTTON_A,     PadButton::PAD_BUTTON_B,
+    PadButton::PAD_BUTTON_X,    PadButton::PAD_BUTTON_Y,     PadButton::PAD_BUTTON_START,
+};
+
+Joycons::Joycons() {
+    if (joycon_handle != nullptr) {
+        return;
+    }
+    LOG_INFO(Input, "JC Adapter Initialization started");
+
+    const int init_res = hid_init();
+    if (init_res ==0) {
+        Setup();
+    } else {
+        LOG_ERROR(Input, "HIDapi could not be initialized. failed with error = {}", init_res);
+    }
+}
+
+void Joycons::Set_rumble(f32 amp_high, f32 amp_low, f32 freq_high, f32 freq_low) {
+    hdRumble.freq_high = std::clamp(freq_high, 81.75177f, 1252.572266f);
+    hdRumble.freq_low = std::clamp(freq_low, 40.875885f, 626.286133f);
+    hdRumble.amp_high = std::clamp(amp_high, 0.0f, 1.0f);
+    hdRumble.amp_low = std::clamp(amp_low, 0.0f, 1.0f);
+    SendRumble();
+}
+void Joycons::SendRumble() {
+    u8* buffer = new u8[49];
+    for (int i = 0; i < 49; ++i) {
+        buffer[i] = 0;
+    }
+    buffer[0] = 0x10;
+    buffer[1] = get_counter();
+    if (hdRumble.amp_high == 0) {
+        buffer[2] = 0x00;
+        buffer[3] = 0x01;
+        buffer[4] = 0x40;
+        buffer[5] = 0x40;
+    } else {
+        u16 encoded_hf = static_cast<u16>(roundf(128 * log2f(hdRumble.freq_high * 0.1f)) - 0x180);
+        u8 encoded_lf = static_cast<u8>(roundf(32 * log2f(hdRumble.freq_low * 0.1f)) - 0x40);
+        u8 encoded_hamp = 0;
+
+        if (hdRumble.amp_high < 0.01182) {
+            encoded_hamp = static_cast<u8>(roundf(pow(hdRumble.amp_high, 1.7f) * 7561));
+        } else if (hdRumble.amp_high < 0.11249) {
+            encoded_hamp = static_cast<u8>(roundf((log(hdRumble.amp_high) * 11.556f) + 55.3f));
+        } else if (hdRumble.amp_high < 0.22498) {
+            encoded_hamp = static_cast<u8>(roundf((log2f(hdRumble.amp_high) * 32) + 131));
+        } else {
+            encoded_hamp = static_cast<u8>(roundf((log2f(hdRumble.amp_high) * 64) + 200));
+        }
+
+        // encoded_lamp should use amp_low data, for now sticking to documentation
+        u8 encoded_lamp = encoded_hamp >> 1;
```

Comment:
```suggestion
        const u8 encoded_lamp = encoded_hamp >> 1;
```

This and most of the assignments below also need an explicit cast to `u8`, since operations on types smaller than `int` promote to `int`, so this will cause truncation warnings to occur. One solution that may be less noisy is to make the involved variables `u32`s and cast them at the end when they're being stored into the buffer.